### PR TITLE
Increase vanilla mushroom spread limit

### PR DESCRIPTION
**Pull request by GrayR0ot**
This update introduces a mushroom configuration to the Paper world, providing users with the ability to bypass the standard vanilla spread limit. This customization allows users to exert more control over mushroom spreading, potentially leading to increased mushroom growth within the designated area.

### CODE REVIEW

File: patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
++++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+@@ -354,6 +354,12 @@ public class WorldConfiguration extends ConfigurationPart {
+         public boolean portalSearchVanillaDimensionScaling = true;
+         public boolean disableTeleportationSuffocationCheck = false;
+         public IntOr.Disabled netherCeilingVoidDamageHeight = IntOr.Disabled.DISABLED;
++
++        public Mushroom mushroom;
++
++        public class Mushroom extends ConfigurationPart {
```
Comment (lynxplay) [inline patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch:18]: New configuration options should be added to the paper configuration patch, not the patch consuming the config option.

File: patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
++            public int spreadLimit = 5;
```
Comment (lynxplay) [inline patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch:19]: I think `IntOr.Default` would fit a lot better here.

File: patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+diff --git a/src/main/java/net/minecraft/world/level/block/MushroomBlock.java b/src/main/java/net/minecraft/world/level/block/MushroomBlock.java
+index 5238b23cd3bca21e2b14c9be15699b95ad267e24..7294d268a04486952389282a0d27f2737f302bd1 100644
+--- a/src/main/java/net/minecraft/world/level/block/MushroomBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/MushroomBlock.java
+@@ -40,10 +40,17 @@ public class MushroomBlock extends BushBlock implements BonemealableBlock {
+     @Override
+     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+         if (random.nextFloat() < (world.spigotConfig.mushroomModifier / (100.0f * 25))) { // Spigot - SPIGOT-7159: Better modifier resolution
+-            int i = 5;
++            io.papermc.paper.configuration.WorldConfiguration.Environment.Mushroom mushroomSettings = world.paperConfig().environment.mushroom; // Paper - Increase vanilla mushroom spread
```
Comment (lynxplay) [inline patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch:33]: What is the point of this ? This variable is never used again.

File: patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
++
+             boolean flag = true;
+             Iterator iterator = BlockPos.betweenClosed(pos.offset(-4, -1, -4), pos.offset(4, 1, 4)).iterator();
+ 
++            int mushroomSpreadLimit = world.paperConfig().environment.mushroom.spreadLimit;
++            if(mushroomSpreadLimit == -1) {
++                mushroomSpreadLimit = 9 * 9 * 3; // Paper - 9x9x3 Vanilla SHAPE sum
++            }
++            int i = mushroomSpreadLimit;
```
Comment (lynxplay) [inline patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch:42]: Why is the definition of i moved below the iterator and flag creation ? This just seems like added diff for no reason.

File: patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch [hunk 5]
```diff
@@ -0,0 +1,46 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?L=C3=A9o=20Moriceau?= <thisis41@protonmail.com>
```
Comment (lynxplay) [inline patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch:2]: Are you sure this is the correct name of the author ?

**Review by lynxplay — CHANGES_REQUESTED** [review]
Welcome to paper :tada: 
Thank you for your first PR! A few notes on the PR in general, tho generally I like the idea.

File: patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch (hunk 3 above)
Comment (GrayR0ot) [reply patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch:33]: This is a copy paste mistake, I'll remove it

File: patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch (hunk 5 above)
Comment (GrayR0ot) [reply patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch:2]: UTF-8 issue, i'll change

File: patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch (hunk 1 above)
Comment (GrayR0ot) [reply patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch:18]: Do you mean I should do a patch with the config, & then another patch that will use it ?

Comment (lynxplay) [reply patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch:18]: The diff in question here should live in the https://github.com/PaperMC/Paper/blob/master/patches/server/0005-Paper-config-files.patch patch.

Comment (GrayR0ot) [reply patches/server/1001-Increase-vanilla-mushroom-spread-limit.patch:18]: oh ok, thanks

**Pull request closed without merging**